lib/Optimizer/Transforms/ApplyControlNegations.cpp

```diff
@@ -0,0 +1,104 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "PassDetails.h"
+#include "cudaq/Optimizer/Builder/Factory.h"
+#include "cudaq/Optimizer/Dialect/CC/CCOps.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeOps.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "cudaq/Todo.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace cudaq::opt {
+#define GEN_PASS_DEF_APPLYCONTROLNEGATIONS
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+using namespace mlir;
+
+/// Replace any operations with negative controls with the same
+/// operation with negative controls and the addition of X operations
+/// on each control qubit before and after the operation.
+template <typename Op>
+class ReplaceNegativeControl : public OpRewritePattern<Op> {
+public:
+  using OpRewritePattern<Op>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(Op op,
+                                PatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+    auto negations = op.getNegatedQubitControls();
+    if (!negations.has_value())
+      return failure();
+
+    for (auto negationIter : llvm::enumerate(negations.value()))
+      if (negationIter.value())
+        rewriter.create<quake::XOp>(
+            loc, ValueRange(),
+            ValueRange{op.getControls()[negationIter.index()]});
```

Comment:
I'm assuming this will work fine if the control argument is type `!quake.veq`.

Reply:
I ignored that case since we don't expose it in the language (we don't have a way to negate `!` a `qvector` or `qview`. Am I wrong to ignore that? Is it possible to have a negation attribute on an operation that takes mixed ref / veq types for the controls?

Reply:
You are correct. Thanks, Alex.